src/Pages/Dashboard/Sensor.jsx

```diff
@@ -21,15 +40,68 @@ class Sensor extends PureComponent {
     this.getDatabase(this.props.sensorId);
   };
 
+  changeRange = (value) => {
+    this.setState({ selected: value });
+    let offset;
+    offset = timeOffsetMap[value];
+    let date = new Date();
+    let now = date.getTime();
+    let d = now - offset;
+    this.timestamp = Math.floor(d / 1000);
+    this.nowstamp = Math.floor(now / 1000);
+    this.setState({ range: this.timestamp });
+    this.getDatabase(this.props.sensorId);
+  };
+
+  handleStartDateTimeChange = (startDate) => {
+    this.setState({ start: startDate });
+    let startTime = startDate.getTime();
+    this.timestamp = Math.floor(startTime / 1000);
+    this.setState({ range: this.timestamp });
+    this.getDatabase(this.props.sensorId);
+  };
+
+  handleEndDateTimeChange = (endDate) => {
+    this.setState({ end: endDate });
+    let endTime = endDate.getTime();
+    this.nowstamp = Math.floor(endTime / 1000);
+    this.setState({ range: this.nowstamp });
+    this.getDatabase(this.props.sensorId);
+  };
+
   getDatabase = sensorId => {
     this.db = this.props.firebase.database().ref(sensorId);
     let temp = [];
-    // Use .once() to make it call less data
-    this.db.limitToLast(30).once("value", e => {
-      for (let i in e.val()) {
-        temp.push(e.val()[i]);
-      }
-    });
+    let tempHistory = [];
+
+    this.db.orderByKey()
+      .startAt(this.timestamp.toString())
+      .endAt(this.nowstamp.toString())
+      .once("value", e => {
+        for (let i in e.val()) {
+          tempHistory.push(e.val()[i]);
+          // temp.push(e.val()[i]); //removed so that main data page shows only live data
+        }
+        this.setState({
+          historyData: tempHistory
+        });
+      });
+
+    this.db.orderByKey()
+      .startAt(this.liveDataStamp.toString())
+      .endAt(this.nowstamp.toString())
+      .once("value", e => {
+        for (let i in e.val()) {
+          temp.push(e.val()[i]);
+          // temp.push(e.val()[i]); //removed so that main data page shows only live data
+        }
+        temp = temp.slice();
+        if (temp.length >= 30) temp.shift();
+        this.setState({
+          data: temp
+        });
+      });
+
```

Comment:
I think a better approach to this would be to have two different branches in terms of logic here: a path that controls the data from `historyData`, and another that controls `data` (you could leave it as the existing logic). 

My first thought would be to have something that sets up `this.db` within the component so that it holds the reference. Then, you can have some function that sets up the live data listener, and another that handles your range changes. 

```javascript
componentDidMount = () => {
    this.initDatabase(this.props.sensorId);
    this.initLiveData();
  };

  changeRange = (value) => {
    this.setState({ selected: value });
    let offset;
    offset = timeOffsetMap[value];
    let date = new Date();
    let now = date.getTime();
    let d = now - offset;
    this.timestamp = Math.floor(d / 1000);
    this.nowstamp = Math.floor(now / 1000);
    this.setState({ range: this.timestamp });
    this.getHistoricalData();
  };

  handleStartDateTimeChange = (startDate) => {
    this.setState({ start: startDate });
    let startTime = startDate.getTime();
    this.timestamp = Math.floor(startTime / 1000);
    this.setState({ range: this.timestamp });
    this.getHistoricalData();
  };

  handleEndDateTimeChange = (endDate) => {
    this.setState({ end: endDate });
    let endTime = endDate.getTime();
    this.nowstamp = Math.floor(endTime / 1000);
    this.setState({ range: this.nowstamp });
    this.getHistoricalData();
  };

  initDatabase = sensorId => {
    this.db = this.props.firebase.database().ref(sensorId);
  }

  initLiveData = () => {
    let temp = [];

    this.db.orderByKey()
    .startAt(this.liveDataStamp.toString())
    .endAt(this.nowstamp.toString())
    .once("value", e => {
      for (let i in e.val()) {
        temp.push(e.val()[i]);
        // temp.push(e.val()[i]); //removed so that main data page shows only live data
      }
      temp = temp.slice();
      if (temp.length >= 30) temp.shift();
      this.setState({
        data: temp
      });
    });

    this.db.limitToLast(1).on("child_added", e => {
      temp = temp.slice();
      if (temp.length >= 30) temp.shift();
      temp.push(e.val());
      this.setState({
        data: temp
      });
    });
  }

  getHistoricalData = () => {
    let tempHistory = [];

    this.db.orderByKey()
      .startAt(this.timestamp.toString())
      .endAt(this.nowstamp.toString())
      .once("value", e => {
        for (let i in e.val()) {
          tempHistory.push(e.val()[i]);
          // temp.push(e.val()[i]); //removed so that main data page shows only live data
        }
        this.setState({
          historyData: tempHistory
        });
      });
  }
```

Something like that, for example. The names could probably use work but the idea is there.